ntoskrnl/lpc/send.c

```diff
@@ -830,9 +976,128 @@ NtRequestWaitReplyPort(IN HANDLE PortHandle,
     /* Check if this is a callback */
     if (Callback)
     {
-        /* FIXME: TODO */
-        Semaphore = NULL; // we'd use the Thread Semaphore here
-        ASSERT(FALSE);
+        /* Check if we have valid request message identifier */
+        if (!LpcRequest->MessageId)
+        {
+            LpcpFreeToPortZone(Message, 0);
+            return STATUS_INVALID_PARAMETER;
+        }
+        Status = PsLookupProcessThreadByCid(&LpcRequest->ClientId,
+                                            NULL,
+                                            &WakeupThread);
+        if (!NT_SUCCESS(Status))
+        {
+            LpcpFreeToPortZone(Message, 0);
+            ObDereferenceObject(Port);
+            return Status;
+        }
+
+        /* Acquire the LPC lock */
+        KeAcquireGuardedMutex(&LpcpLock);
+
+        /* Make sure this is the reply the thread is waiting for */
+        if ((WakeupThread->LpcReplyMessageId != LpcRequest->MessageId) ||
+            ((LpcpGetMessageFromThread(WakeupThread)) &&
+            (LpcpGetMessageType(&LpcpGetMessageFromThread(WakeupThread)->Request)
+                != LPC_REQUEST)) ||
+            (!LpcpValidateClientPort(WakeupThread, Port)))
+        {
+            /* It isn't, fail */
+            LpcpFreeToPortZone(Message, LPCP_LOCK_HELD | LPCP_LOCK_RELEASE);
+            ObDereferenceObject(WakeupThread);
+            ObDereferenceObject(Port);
+            return STATUS_REPLY_MESSAGE_MISMATCH;
+        }
+
+        _SEH2_TRY
+        {
+            LpcpMoveMessage(&Message->Request,
+                            &CapturedLpcRequest,
+                            LpcRequest + 1,
+                            MessageType,
+                            &Thread->Cid);
+            Status = LpcpVerifyMessageDataInfo(LpcRequest, &NumberOfDataEntries);
+            if (!NT_SUCCESS(Status))
+            {
+                DPRINT1("LpcpVerifyMessageDataInfo failed\n");
+                LpcpFreeToPortZone(Message, LPCP_LOCK_HELD | LPCP_LOCK_RELEASE);
+                ObDereferenceObject(WakeupThread);
+                ObDereferenceObject(Port);
+                _SEH2_YIELD(return Status);
+            }
+        }
+        _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+        {
+            DPRINT1("Got exception\n");
```

Comment:
Useless unless you also print the exception code.

Reply:
Right.